Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/WatchVariablePicker.tsx

```diff
@@ -0,0 +1,263 @@
+  }, [payload]);
+
+  const onHideContextualMenu = () => {
+    setShowContextualMenu(false);
+  };
+
+  const getContextualMenuItems = (): IContextualMenuItem[] => {
+    const { root } = propertyTreeConfig;
+    const { nodes, levels, paths } = getAllNodes<PropertyItem>(root, {
+      expanded: propertyTreeExpanded,
+      skipRoot: true,
+    });
+
+    const onToggleExpand = (itemId: string, expanded: boolean) => {
+      setPropertyTreeExpanded({ ...propertyTreeExpanded, [itemId]: expanded });
+    };
+
+    return nodes.map((node) => ({
+      key: node.id,
+      text: node.name,
+      secondaryText: paths[node.id],
+      onClick: (event) => {
+        if (node.children.length) {
+          event?.preventDefault();
+          onToggleExpand(node.id, !propertyTreeExpanded[node.id]);
+        } else {
+          if (currentProjectId) {
+            const path = paths[node.id];
+            setQuery(path);
+            event?.preventDefault();
+            setErrorMessage('');
+            setWatchedVariables(currentProjectId, { ...watchedVariables, [variableId]: path });
+            onHideContextualMenu();
+          }
+        }
+      },
+      data: {
+        node,
+        path: paths[node.id],
+        level: levels[node.id] - 1,
+        onToggleExpand,
+      },
+    }));
+  };
+
+  const onShowContextualMenu = (event: FocusEvent<HTMLInputElement>) => {
+    event.preventDefault();
+    setShowContextualMenu(true);
+  };
+
+  const flatPropertyListItems = React.useMemo(() => {
+    const { root } = propertyTreeConfig;
+    const { nodes, paths } = getAllNodes<PropertyItem>(root, { skipRoot: true });
+
+    return nodes.map((node) => ({
+      text: node.id,
+      key: node.id,
+      secondaryText: paths[node.id],
+      onClick: (event) => {
+        event?.preventDefault();
+        if (currentProjectId) {
+          const path = paths[node.id];
+          setErrorMessage('');
+          setWatchedVariables(currentProjectId, { ...watchedVariables, [variableId]: path });
+          onHideContextualMenu();
+        }
+      },
+      data: {
+        node,
+        path: paths[node.id],
+        level: 0,
+      },
+    })) as IContextualMenuItem[];
+  }, [currentProjectId, payload, propertyTreeConfig, variableId, watchedVariables]);
+
+  const getFilterPredicate = useCallback((q: string) => {
+    return (item: IContextualMenuItem) =>
+      item.data.node.children.length === 0 && item.secondaryText?.toLowerCase().indexOf(q.toLowerCase()) !== -1;
+  }, []);
+
+  const menuItems = useMemo(getContextualMenuItems, [payload, propertyTreeExpanded]);
+
+  useEffect(() => setItems(menuItems), [menuItems]);
+
+  const handleDebouncedSearch: () => void = useCallback(
+    debounce(() => {
+      if (query) {
+        const searchableItems = flatPropertyListItems;
+
+        const predicate = getFilterPredicate(query);
+
+        const filteredItems = searchableItems.filter(predicate);
```

Comment:
`.filter()` makes a new array anyway, so there's no need to reassign flatPropertyListItems to searchableItems in here. I think just `const filteredItems = flatPropertyListItems.filter(predicate)` ought to work.